### PR TITLE
Re-4236: Create kong redirect plugin

### PR DESCRIPTION
**Pull request by abdelhegazi**

### CODE REVIEW

File: kong/plugins/url-redirect/handler.lua [hunk 1]
```diff
@@ -0,0 +1,28 @@
+local BasePlugin = require "kong.plugins.base_plugin"
```
Comment (mattoddie) [inline kong/plugins/url-redirect/handler.lua:1]: This file should be in `kong/plugins/redirect` to match the plugin name (redirect vs url-redirect)

File: kong/plugins/url-redirect/schema.lua [hunk 2]
```diff
@@ -0,0 +1,20 @@
+return {
```
Comment (mattoddie) [inline kong/plugins/url-redirect/schema.lua:1]: Can we update this to closer match the schema format in the template https://github.com/Kong/kong-plugin/blob/master/kong/plugins/myplugin/schema.lua

File: kong/plugins/url-redirect/handler.lua (hunk 1 above)
Comment (mattoddie) [reply kong/plugins/url-redirect/handler.lua:1]: Also can the file closer match the template format for handler https://github.com/Kong/kong-plugin/blob/master/kong/plugins/myplugin/handler.lua

File: kong-plugin-redirect-0.1.0-1.rockspec [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+}
+dependencies = {
+  "lua >= 5.4",
+}
+build = {
+  type = "builtin",
+  modules = {
+    ["kong.plugins.url-redirect.handler"] = "./kong/plugins/url-redirect/handler.lua",
+    ["kong.plugins.url-redirect.schema"] = "./kong/plugins/url-redirect/schema.lua",
+    ["kong.plugins.url-redirect.utils"] = "./kong/plugins/url-redirect/utils.lua"
```
Comment (mattoddie) [inline kong-plugin-redirect-0.1.0-1.rockspec:19]: We don't have `utils` package

**Review by mattoddie — CHANGES_REQUESTED** [review]

File: kong/plugins/url-redirect/handler.lua [hunk 4]
```diff
@@ -0,0 +1,25 @@
+local plugin = {
+    PRIORITY = 3000,
+    VERSION = "0.1.0",
+  }
+
+RedirectHandler.PRIORITY = plugin.PRIORITY
```
Comment (mattoddie) [inline kong/plugins/url-redirect/handler.lua:6]: No need for this line

File: kong/plugins/url-redirect/handler.lua [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+RedirectHandler.VERSION = plugin.VERSION
```
Comment (mattoddie) [inline kong/plugins/url-redirect/handler.lua:7]: No need for this line

File: kong/plugins/url-redirect/handler.lua [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+local BasePlugin = require "kong.plugins.base_plugin"
```
Comment (mattoddie) [inline kong/plugins/url-redirect/handler.lua:8]: No need for this line

File: kong/plugins/url-redirect/handler.lua [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+function RedirectHandler:new()
```
Comment (mattoddie) [inline kong/plugins/url-redirect/handler.lua:10]:
```suggestion
function plugin:new()
```

File: kong/plugins/url-redirect/handler.lua [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    RedirectHandler.super.new(self, "kong-plugin-redirect")
+end
+
+function RedirectHandler:access(conf)
```
Comment (mattoddie) [inline kong/plugins/url-redirect/handler.lua:14]:
```suggestion
function plugin:access(conf)
```

File: kong/plugins/url-redirect/handler.lua [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+    local status_code     = conf.status_code
+    local request_uri     = conf.request_uri
+    local response_uri    = conf.response_uri
+    if ngx.var.request_uri == request_uri then
+        return ngx.redirect(response_uri, status_code)
+    end
+end
+
+return RedirectHandler
```
Comment (mattoddie) [inline kong/plugins/url-redirect/handler.lua:25]:
```suggestion
return plugin
```

File: kong/plugins/url-redirect/schema.lua [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+  fields = {
+    { consumer = typedefs.no_consumer },
+    { protocols = typedefs.protocols_http },
+
+    { config = {
+        type = "record",
+        fields = {
+          { request_uri = typedefs.header_name {
+              required = true,
+              default = "github.com/resolutionlife/kong-plugin-redirect" } },
```
Comment (mattoddie) [inline kong/plugins/url-redirect/schema.lua:15]: The request URI you get in `ngx.var.request_uri` is actually the path

```suggestion
              default = "/resolutionlife/kong-plugin-redirect" } },
```

**Review by mattoddie — CHANGES_REQUESTED** [review]

File: kong/plugins/url-redirect/handler.lua (hunk 6 above)
Comment (abdelhegazi) [reply kong/plugins/url-redirect/handler.lua:5]: I guess we need this `BasePlugin` for the redirect

File: kong/plugins/url-redirect/schema.lua [hunk 11]
```diff
@@ -0,0 +1,27 @@
+local typedefs = require "kong.db.schema.typedefs"
+local PLUGIN_NAME = "kong-plugin-redirect"
```
Comment (mattoddie) [inline kong/plugins/url-redirect/schema.lua:2]: `kong-plugin-redirect` is repo name, not plugin name

```suggestion
local PLUGIN_NAME = "kong-redirect"
```

File: kong/plugins/url-redirect/handler.lua [hunk 12]
```diff
@@ -0,0 +1,16 @@
+local plugin = {
+    PRIORITY = 3000,
+    VERSION = "0.1.0",
+  }
+
+function plugin:access(conf)
+
```
Comment (mattoddie) [inline kong/plugins/url-redirect/handler.lua:7]:
```suggestion
```

File: kong-plugin-redirect-0.1.0-1.rockspec [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+  summary = "A Kong plugin that enables elastic request/response transforming by lua scripts",
+  license = "MIT License",
+}
+dependencies = {
+  "lua >= 5.4",
+}
+build = {
+  type = "builtin",
+  modules = {
+    ["kong.plugins.url-redirect.handler"] = "./kong/plugins/url-redirect/handler.lua",
```
Comment (mattoddie) [inline kong-plugin-redirect-0.1.0-1.rockspec:17]: Can we replace all instances of "url-redirect" with just redirect, including plugin folder name?

File: kong-plugin-redirect-0.1.0-1.rockspec [hunk 14]
```diff
@@ -0,0 +1,21 @@
+package = "kong-plugin-redirect"
```
Comment (mattoddie) [inline kong-plugin-redirect-0.1.0-1.rockspec:1]: I think the plugin template file https://github.com/Kong/kong-plugin/blob/master/kong-plugin-myplugin-0.1.0-1.rockspec is much nicer for all of this as it removes a lot of copy/paste by having variables etc.

File: .github/workflows/test.yaml [hunk 15]
```diff
@@ -0,0 +1,57 @@
+name: "Test"
```
Comment (mattoddie) [inline .github/workflows/test.yaml:1]: Can we remove this file since we have no tests?

**Review by mattoddie — CHANGES_REQUESTED** [review]

**Review by mattoddie — APPROVED** [review]

**Pull request merged**